[SR-610][stdlib] Use SequenceType._initializeTo where possible

lorentey opened the pull request:
(This is an updated version of PR #1084.)

Change the `SequenceType._initializeTo` entry point to support memory safety better,
and use it in more places where `Array` elements are initialized in bulk:
- Conversion of sequences and collections to `Array`
- `<ArrayType> += <SequenceType>`
- `ArrayType.replaceRange` (when used for insertion)

The speedup achieved is highly sensitive to the mood of the optimizer. The most remarkable result is that insertion of an `Array` into another `Array` can be 4-35x faster depending on the element type. Besides `Array`, `ArraySlice` and `ContiguousArray` also have custom `_initializeTo` implementations, but interestingly, only `Array` seems to be meaningfully impacted by these changes. (The behavior of the
optimizer can be highly unintuitive, making attempts at micro-optimization less useful; see SR-653.)

When `_initializeTo` isn't available, prefer `Generator` to looping over index ranges when iterating over elements of a collection. This speeds up the above operations for collections with complex indexes. 
For example, operations on standard hashed collections typically improve by about 10-40%.

lorentey commented:
(This is an updated version of PR #1084.)

lorentey commented:
I attached detailed benchmarking results as [SR-610.xlsx](https://github.com/apple/swift/files/116531/SR-610.xlsx). The benchmark measures execution time of the following operations, for each of `Array<Int>`, `Array<String>`, `Array<(Int, String)>`, `ArraySlice<Int>`, `ArraySlice<(Int, String)>`, `Dictionary<Int, String>`, `Set<Int>`, `Set<String>`, `Deque<Int>`, `Deque<(Int, String)>`, and `Map<Int, String>`:

1. Converting the collection to `Array` via `Array(thing)`
2. Converting the collection to `Array` using `Array(GeneratorSequence(thing.generate()))`
3. Appending the contents of the collection to the end of an `Array` using `+=`
4. Inserting the contents of the collection in the middle of an `Array` using `insertContentsOf`

The `Deque` type is a custom double-ended queue built on a circular buffer. `Map` is an associative collection like `Dictionary`, but it uses `Comparable` keys and it is built on an in-memory b-tree. As a sanity check, the results also contain control measurements marked "ZZZ Control" that do not use modified codepaths, and should measure the same before/after the patch is applied.

The source code for my benchmark utility is at
[SR610Benchmark.zip](https://github.com/apple/swift/files/116543/SR610Benchmark.zip).

lorentey commented on stdlib/public/core/ContiguousArrayBuffer.swift:
```diff
@@ -582,38 +570,18 @@ extension _ContiguousArrayBuffer {
 }
 
 /// This is a fast implementation of _copyToNativeArrayBuffer() for collections.
-///
```
Benchmarking indicates that using `_UnsafePartiallyInitializedContigiousArrayBuffer` does not incur a measurable performance penalty. In fact, `addFrom` somehow compiles into faster code than the simple inline loop that's in the original code.

lorentey commented on stdlib/public/core/ContiguousArrayBuffer.swift:
```diff
@@ -582,38 +570,18 @@ extension _ContiguousArrayBuffer {
-/// _UnsafePartiallyInitializedContiguousArrayBuffer. Since we do not support
-/// ARC loops, the extra retain, release overhead cannot be eliminated which
-/// makes assigning ranges very slow. Once this has been implemented, this code
-/// should be changed to use _UnsafePartiallyInitializedContiguousArrayBuffer.
```
Benchmarking indicates that using `_UnsafePartiallyInitializedContigiousArrayBuffer` does not incur a measurable performance penalty. In fact, `addFrom` somehow compiles into faster code than the simple inline loop that's in the original code.

lorentey commented on stdlib/private/StdlibCollectionUnittest/MinimalCollections.swift.gyb:
```diff
@@ -1341,7 +1341,11 @@ public struct Defaulted${traversal}RangeReplaceableSlice<Element>
   }
 
   public func generate() -> MinimalGenerator<Element> {
-    return MinimalGenerator(Array(self))
```
The original code uses an API requiring iteration in the implementation of the iteration API itself; this leads to infinite recursion with the PR's changes.

lorentey commented on validation-test/stdlib/ArrayNew.swift.gyb:
```diff
@@ -1610,9 +1610,8 @@ final class EvilCollection : CollectionType {
 
-  let message = step < 0 && evilBoundsCheck
-    ? evilBoundsError
-    : "invalid CollectionType: count differed in successive traversals"
```
Usage of the new `expectStableCollection` function ensures that evil collections get consistent fatal error messages.

lorentey commented on stdlib/public/core/Collection.swift:
```diff
@@ -589,28 +589,6 @@ extension CollectionType
-  where Self : _ArrayType, Self.Element == Self.Generator.Element {
-  // A fast implementation for when you are backed by a contiguous array.
-  public func _initializeTo(ptr: UnsafeMutablePointer<Generator.Element>)
-    -> UnsafeMutablePointer<Generator.Element> {
```
This overload proved ineffective for e.g. `ArraySlice`, which has a nil `_baseAddressIfContiguous`. I instead added an `_initializeTo` implementation
in `Arrays.swift.gyb` that uses `_buffer._uninitializedCopy`.

gribozavr commented:
@swift-ci Please test

jckarter commented:
@aschwaighofer and/or @nadavrot will probably want to look at this. I know we've gone back and forth on some of these micro optimizations as the high level optimizations passes have improved.

nadavrot commented:
@lorentey It looks like your commit squashed a bunch of different changes (to initializeTo,  to stdlibunittests, etc). Can you split the commit into multiple patches that can be reviewed independently?

lorentey commented:
@nadavrot Absolutely! Would these three phases be easier to review?

- PR 1: Update `_initializeTo` to the new signature everywhere.
- PR 2: Use the new `_initializeTo` in more operations.
- PR 3: Prefer `generate()` to indexing.

lorentey commented:
@nadavrot I split the PR into three distinct commits. I hope it will be easier to review now!

lorentey commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test

lorentey commented:
I'd like to move this forward; is there anything I can do to help?

gribozavr commented:
@lorentey I'm sorry for not responding!  Unfortunately I was busy.  I will try to review this PR.

One thing that you could do is to run the performance testsuite that was recently added to the repository (built binary at `bin/Benchmark_O`).  You'd need to build with `-R --no-swift-stdlib-assertions` to get meaningful benchmark results.

lorentey commented:
No worries! This isn't urgent at all, I just don't want to forget about it.

I ran `utils/build-script -R --no-swift-stdlib-assertions --benchmark` a couple of times, collecting the `-O` results in a spreadsheet ([PDF](https://github.com/apple/swift/files/131832/Comparison.pdf), [Excel](https://github.com/apple/swift/files/131833/Comparison.xlsx)).

The proposed change seems to be a slight overall improvement (1.5% total) in optimized builds. 17 out of  78 tests show larger than 3% improvement, while 4 others (`ClassArrayGetter`, `RangeAssignment`, `RGBHistogram` and `Sim2DArray`) have at least the same delta in the negative direction. Of the latter, `RGBHistogram` is particularly interesting, because for some reason it exhibits a whopping 35% slowdown; the three others are slower by a relatively boring 6-8%.

I haven't analyzed `-Onone` results yet, but anectodally I did notice major (30-100%) slowdowns for `MapReduce`, `NopDeinit`, and `RangeAssignment`. (The rest of the tests seem to be a mixed bag when optimization is off.)

I will look into the tests to see if I can explain these outliers and hopefully fix them.

gribozavr commented:
It would be good if you could look into RGBHistogram.  The first thing to do is to increase the number of samples to ensure that the difference is not caused by noise: 

```
./bin/Benchmark_O --num-samples=10 RGBHistogram
```

nadavrot commented:
I don’t think that it is very likely that a change to SequenceType improved the performance of the Calculator test by so much. I recommend running the benchmark again with a higher number of iterations. 

> On Feb 15, 2016, at 7:47 PM, Károly Lőrentey <notifications@github.com> wrote:
> 
> No worries! This isn't urgent at all, I just don't want to forget about it.
> 
> I ran utils/build-script -R --no-swift-stdlib-assertions --benchmark a couple of times, collecting the -O results in a spreadsheet (PDF <https://github.com/apple/swift/files/131832/Comparison.pdf>, Excel <https://github.com/apple/swift/files/131833/Comparison.xlsx>).
> 
> The proposed change seems to be a slight overall improvement (1.5% total) in optimized builds. 17 out of 78 tests show larger than 3% improvement, while 4 others (ClassArrayGetter, RangeAssignment, RGBHistogram and Sim2DArray) have at least the same delta in the negative direction. Of the latter, RGBHistogram is particularly interesting, because for some reason it exhibits a whopping 35% slowdown; the three others are slower by a relatively boring 6-8%.
> 
> I haven't analyzed -Onone results yet, but anectodally I did notice major (30-100%) slowdowns for MapReduce, NopDeinit, and RangeAssignment. (The rest of the tests seem to be a mixed bag when optimization is off.)
> 
> I will look into the tests to see if I can explain these outliers and hopefully fix them.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/apple/swift/pull/1194#issuecomment-184504629>.
>

lorentey commented:
@gribozavr @nadavrot Increasing the number of samples & iterations eliminates the slowdown for `ClassArrayGetter`, but keeps it for the rest of the big losers, including `RGBHistogram`.

The `Calculator` result is reliably repeatable; this is perplexing. I'll add markers to code paths I touched to see if they're somehow involved. 

`RGBHistogram` calls `sort` on a dictionary of integers, which seems like a nice candidate for me to start looking for regressions.

Raw results for 100 samples and 100 iterations are below:

#### Before

`./build/master/swift-macosx-x86_64/bin/Benchmark_O --num-samples=100 --num-iters=100 RGBHistogram Calculator ClassArrayGetter RangeAssignment Sim2DArray`

#|TEST|SAMPLES|MIN(us)|MAX(us)|MEAN(us)|SD(us)|MEDIAN(us)
--- | --- | --- | --- | --- | --- | --- | ---
18|Calculator|100|44|85|46|6|45
21|ClassArrayGetter|100|25|41|27|3|26
56|RGBHistogram|100|3147|3411|3251|60|3246
57|RangeAssignment|100|332|471|351|20|347
64|Sim2DArray|100|777|902|809|29|802
Totals|5||4325|4910|4484|0|0

#### After

`./build/sr610/swift-macosx-x86_64/bin/Benchmark_O --num-samples=100 --num-iters=100 RGBHistogram Calculator ClassArrayGetter RangeAssignment Sim2DArray`

#|TEST|SAMPLES|MIN(us)|MAX(us)|MEAN(us)|SD(us)|MEDIAN(us)
--- | --- | --- | --- | --- | --- | --- | ---
18|Calculator|100|38|71|41|5|40
21|ClassArrayGetter|100|25|40|26|2|25
56|RGBHistogram|100|4310|4737|4442|82|4427
57|RangeAssignment|100|352|427|371|17|368
64|Sim2DArray|100|786|964|830|40|828
Totals|5||5511|6239|5710|0|0

lorentey commented:
Update on `Calculator`: taking the `master` branch and adding a `print()` call to `CollectionType.map` in `Collections.swift` is enough to reproduce the speedup. This is true even though the `map` function is not actually called by the benchmark case. Hm.

This is with commit 2c0fd720 from yesterday; I'm going to pull the latest `master` and retry.

lorentey commented:
I looked into `RGBHistogram`. The slowdown is due to the array conversion in `Dictionary.sort`, and can be entirely attributed to a single inlining decision: when I add `@inline(__always)` on `DictionaryGenerator.next()`, the slowdown disappears.

This suggests the same inlining decision are also responsible for the 10-40% *speedup* I measured in my own `Dictionary` benchmarks, which were all using native dictionaries.

I still expect the new code to behave meaningfully better for bridged `NSDictionary` instances, where the generator is able to use fast enumeration, while creating indexes is slow.

gottesmm commented:
> On Feb 16, 2016, at 5:23 AM, Károly Lőrentey <notifications@github.com> wrote:
> 
> @gribozavr <https://github.com/gribozavr> @nadavrot <https://github.com/nadavrot> Increasing the number of samples & iterations eliminates the slowdown for ClassArrayGetter, but keeps it for the rest of the big losers, including RGBHistogram.
> 
> 

Just an FYI, you probably don't want to touch the iteration count. What you want to change is the sample count.
> The Calculator result is reliably repeatable; this is perplexing. I'll add markers to code paths I touched to see if they're somehow involved.
> 
> RGBHistogram calls sort on a dictionary of integers, which seems like a nice candidate for me to start looking for regressions.
> 
> Raw results for 100 samples and 100 iterations are below:
> 
> Before
> 
> ./build/master/swift-macosx-x86_64/bin/Benchmark_O --num-samples=100 --num-iters=100 RGBHistogram Calculator ClassArrayGetter RangeAssignment Sim2DArray
> 
> #	TEST	SAMPLES	MIN(us)	MAX(us)	MEAN(us)	SD(us)	MEDIAN(us)
> 18	Calculator	100	44	85	46	6	45
> 21	ClassArrayGetter	100	25	41	27	3	26
> 56	RGBHistogram	100	3147	3411	3251	60	3246
> 57	RangeAssignment	100	332	471	351	20	347
> 64	Sim2DArray	100	777	902	809	29	802
> Totals	5		4325	4910	4484	0	0
> After
> 
> ./build/sr610/swift-macosx-x86_64/bin/Benchmark_O --num-samples=100 --num-iters=100 RGBHistogram Calculator ClassArrayGetter RangeAssignment Sim2DArray
> 
> #	TEST	SAMPLES	MIN(us)	MAX(us)	MEAN(us)	SD(us)	MEDIAN(us)
> 18	Calculator	100	38	71	41	5	40
> 21	ClassArrayGetter	100	25	40	26	2	25
> 56	RGBHistogram	100	4310	4737	4442	82	4427
> 57	RangeAssignment	100	352	427	371	17	368
> 64	Sim2DArray	100	786	964	830	40	828
> Totals	5		5511	6239	5710	0	0
> —
> Reply to this email directly or view it on GitHub <https://github.com/apple/swift/pull/1194#issuecomment-184681596>.
>

gottesmm commented:
> On Feb 17, 2016, at 5:54 AM, Károly Lőrentey <notifications@github.com> wrote:
> 
> I looked into RGBHistogram. The slowdown is due to the array conversion in Dictionary.sort, and can be entirely attributed to a single inlining decision: when I add @inline(__always) on DictionaryGenerator.next(), the slowdown disappears.
> 
> This suggests the same inlining decision are also responsible for the 10-40% speedup I measured in my own Dictionary benchmarks, which were all using native dictionaries.
> 
> I still expect the new code to behave meaningfully better for bridged NSDictionary instances, where the generator is able to use fast enumeration, while creating indexes is slow.
> 
> 

Did you look in instruments at the perf slow down? Is it an ARC issue?
> —
> Reply to this email directly or view it on GitHub <https://github.com/apple/swift/pull/1194#issuecomment-185213079>.
>

gribozavr commented:
CC @rudkx for the inlining difference.  If this is such a sensitive method, do we need to mark it always-inline?

nadavrot commented:
+ @eeckstein 

The standard library is special (it is compute intensive and we are willing to invest in hand optimizing it) so it is okay to add inlining attributes to force specific behavior.

rudkx commented:
Agree it's fine to add the @inline(__always) if it helps here. @eeckstein is looking at inlining now, to it's also worth looking at whether we can improve the inliner to inline this case without forcing it.

dabrahams commented:
@lorentey Thanks for all your work on this.  Could you update it so it merges cleanly and add the `@inline(__always)` annotation?  If you can do it, running the performance tests again would be great too.

@nadavrot I plan to accept this PR when it merges cleanly unless the performance team wants to step in and make specific requests.

lorentey commented:
Of course! I'll update the branch & re-run the tests in a couple of hours.

lorentey commented:
@dabrahams I fixed the issues in your review, and rebased the branch on a more recent master, so it merges cleanly now. I added a new commit for `@inline(__always)`.

I'm going to run the benchmarks next.

gribozavr commented:
@swift-ci Please test

lorentey commented:
The benchmark once again contains an interesting surprise, with `StringWithCString` becoming 2x slower. I'm investigating!


````
  # TEST                      BEST_OLD_MIN(μs) BEST_NEW_MIN(μs)     DELTA    %DELTA   SPEEDUP
  1 AngryPhonebook                        2772             2744       -28     -1.0%     1.01x (?)
  2 Array2D                               2956             2909       -47     -1.6%     1.02x (?)
  3 ArrayAppend                            711              700       -11     -1.5%     1.02x (?)
  4 ArrayAppendReserved                    483              477        -6     -1.2%     1.01x (?)
  5 ArrayInClass                           100              100        +0     +0.0%     1.00x
  6 ArrayLiteral                          1002             1000        -2     -0.2%     1.00x
  7 ArrayOfGenericPOD                      207              203        -4     -1.9%     1.02x (?)
  8 ArrayOfGenericRef                     3597             3494      -103     -2.9%     1.03x (?)
  9 ArrayOfPOD                             170              168        -2     -1.2%     1.01x (?)
 10 ArrayOfRef                            3428             3437        +9     +0.3%     1.00x (?)
 11 ArraySubscript                        1423             1375       -48     -3.4%     1.03x (?)
 12 ArrayValueProp                           8                8        +0     +0.0%     1.00x
 13 ArrayValueProp2                          8                8        +0     +0.0%     1.00x
 14 ArrayValueProp3                          8                8        +0     +0.0%     1.00x
 15 ArrayValueProp4                          8                8        +0     +0.0%     1.00x
 16 BitCount                                17               18        +1     +5.9%     0.94x
 17 ByteSwap                                 1                1        +0     +0.0%     1.00x
 18 Calculator                              39               40        +1     +2.6%     0.97x (?)
 19 CaptureProp                           5663             5300      -363     -6.4%     1.07x
 20 Chars                                  874              878        +4     +0.5%     1.00x (?)
 21 ClassArrayGetter                        12               12        +0     +0.0%     1.00x
 22 DeadArray                              169              195       +26    +15.4%     0.87x
 23 Dictionary                            1902             1873       -29     -1.5%     1.02x (?)
 24 Dictionary2                           3541             3498       -43     -1.2%     1.01x (?)
 25 Dictionary3                           1291             1285        -6     -0.5%     1.00x (?)
 26 DictionaryBridge                      5161             4880      -281     -5.4%     1.06x (?)
 27 DictionaryLiteral                     2807             2769       -38     -1.4%     1.01x (?)
 28 DictionaryRemove                      4941             4632      -309     -6.3%     1.07x (?)
 29 DictionarySwap                        1002             1005        +3     +0.3%     1.00x
 30 ErrorHandling                         2774             2640      -134     -4.8%     1.05x (?)
 31 GlobalClass                              0                0        +0         *         *
 32 Hanoi                                 3525             3512       -13     -0.4%     1.00x (?)
 33 HashTest                              1710             1701        -9     -0.5%     1.01x (?)
 34 Histogram                             1014              730      -284    -28.0%     1.39x
 35 Integrate                              245              240        -5     -2.0%     1.02x (?)
 36 Join                                   497              484       -13     -2.6%     1.03x (?)
 37 LinkedList                           10046             9386      -660     -6.6%     1.07x
 38 MapReduce                              410              406        -4     -1.0%     1.01x (?)
 39 Memset                                 213              207        -6     -2.8%     1.03x (?)
 40 MonteCarloE                          10470            10112      -358     -3.4%     1.04x (?)
 41 MonteCarloPi                         43432            42761      -671     -1.5%     1.02x (?)
 42 NSDictionaryCastToSwift              11179            10932      -247     -2.2%     1.02x (?)
 43 NSError                                360              356        -4     -1.1%     1.01x (?)
 44 NSStringConversion                     620              621        +1     +0.2%     1.00x (?)
 45 NopDeinit                            34843            34247      -596     -1.7%     1.02x
 46 ObjectAllocation                       229              223        -6     -2.6%     1.03x (?)
 47 OpenClose                               67               66        -1     -1.5%     1.02x (?)
 48 Phonebook                            10799            10719       -80     -0.7%     1.01x (?)
 49 PolymorphicCalls                       101              100        -1     -1.0%     1.01x
 50 PopFrontArray                         1938             1909       -29     -1.5%     1.02x (?)
 51 PopFrontArrayGeneric                  1941             1909       -32     -1.6%     1.02x (?)
 52 PopFrontUnsafePointer                10038             9235      -803     -8.0%     1.09x
 53 Prims                                  771              757       -14     -1.8%     1.02x (?)
 54 ProtocolDispatch                      3219             3189       -30     -0.9%     1.01x (?)
 55 ProtocolDispatch2                      145              144        -1     -0.7%     1.01x (?)
 56 RC4                                    222              219        -3     -1.4%     1.01x (?)
 57 RGBHistogram                          3005             2990       -15     -0.5%     1.01x (?)
 58 RangeAssignment                        330              346       +16     +4.8%     0.95x (?)
 59 RecursiveOwnedParameter               1954             1931       -23     -1.2%     1.01x (?)
 60 SetExclusiveOr                        3303             3407      +104     +3.1%     0.97x (?)
 61 SetIntersect                           398              392        -6     -1.5%     1.02x (?)
 62 SetIsSubsetOf                          565              557        -8     -1.4%     1.01x (?)
 63 SetUnion                              3048             3124       +76     +2.5%     0.98x (?)
 64 SevenBoom                             1352             1343        -9     -0.7%     1.01x (?)
 65 Sim2DArray                             785              783        -2     -0.3%     1.00x (?)
 66 SortLettersInPlace                    1008             1001        -7     -0.7%     1.01x
 67 SortStrings                           1786             1757       -29     -1.6%     1.02x (?)
 68 StaticArray                           3008             2862      -146     -4.9%     1.05x (?)
 69 StrComplexWalk                        3333             3245       -88     -2.6%     1.03x (?)
 70 StrToInt                              4190             4239       +49     +1.2%     0.99x (?)
 71 StringBuilder                         1270             1245       -25     -2.0%     1.02x (?)
 72 StringInterpolation                  10548            10886      +338     +3.2%     0.97x (?)
 73 StringWalk                            5078             4996       -82     -1.6%     1.02x
 74 StringWithCString                    82462           176535    +94073   +114.1%     0.47x
 75 SuperChars                          633760           574407    -59353     -9.4%     1.10x
    Totals                              958825           988314    +29489     +3.1%     0.97x (?)
 76 TwoSum                                1771             1716       -55     -3.1%     1.03x (?)
 77 TypeFlood                                0                0        +0         *         *
 78 Walsh                                  387              365       -22     -5.7%     1.06x
 79 XorLoop                                375              357       -18     -4.8%     1.05x
````

lorentey commented:
BTW, `DeadArray` becoming 15% slower looks like the `Calculator` case to me. It doesn't look like the test touches on any modified code, does it?

I believe the big win with `Histogram` is due to iteration now being inlined.

lorentey commented:
`StringWithCString` appends an `UTF8View` to a `ContiguousArray`. `UTF8View` is a collection with an `IndexingGenerator`, so my hypothesis is that the slowdown is due to another inlining decision.

This is very interesting, because it shows that for collections where the generator is a trivial wrapper around `index.successor()`, the compiler often fails to inline the wrapper on its own.

(Most of my experience is with tree-based collections where iteration using indices generally cost much more than generators. @nadavrot has asked me to add a benchmark for my B-trees; perhaps it would be useful if I moved that a little higher on my todo list.)

lorentey commented:
Yep, adding `@inline(__always)` to `IndexingGenerator.next()` makes the slowdown disappear:

```
$ ../build/Ninja-ReleaseAssert+stdlib-Release/swift-macosx-x86_64/bin/Benchmark_O StringWithCString --num-samples=25
#,TEST,SAMPLES,MIN(us),MAX(us),MEAN(us),SD(us),MEDIAN(us)
74,StringWithCString,25,82928,94806,87397,3616,86318

Totals,1,82928,94806,87397,0,0
```

lorentey commented:
[My latest results](https://github.com/apple/swift/files/157569/Benchmark_O-20160303202752.txt)
are more promising, but looking closer at the numbers, it looks like the benchmark runner has a serious bug: it sometimes reports minimum values that are higher than the maximum or mean. Here is a snippet from the raw benchmark logs for one of my latest runs:

```
33,HashTest,25,1700,1785,1731,0,1731,10512138
34,Histogram,25,1023,761,756,0,756,3915940
35,Integrate,25,244,253,248,0,248,3910697
```

Note how `Histogram` has a huge "minimum" value at 1023 (matching the +/-40% speedup that comes and goes), while the maximum is 761 and the mean is 756. A number of other measurements are also misreported like this in the same log. Unless I'm misreading these numbers, this may make all results produced by the current version of the benchmark suite highly suspect.

gottesmm commented:
For the room. @lorentey filed https://bugs.swift.org/browse/SR-871 for the min/max issue.

dabrahams commented:
on Thu Mar 03 2016, Károly Lőrentey <notifications-AT-github.com> wrote:

> Yep, adding `@inline(__always)` to `IndexingGenerator.next()` makes the slowdown disappear:

And that sounds like an obviously-good change.
-- 
-Dave

lorentey commented:
I rebased the commits on a version of master that includes the fix for SR-871, and re-run the benchmarks.  Half of the `StringWithCString` slowdown came back on latest master, so I also force-inlined `String.UTF8View.successor()` to bring it down to +6%. The results look okay I think. 

(+11% on `PhoneBook` is new. It uses `Array.append` and `sortInPlace`, neither of which should be affected by any of these changes.)

I'll shortly push an update with the latest code.

````
Comparing master/Benchmark_O-20160304142049.log sr-610/Benchmark_O-20160304161937.log ...
  # TEST                      BEST_OLD_MIN(μs) BEST_NEW_MIN(μs)     DELTA    %DELTA   SPEEDUP
  1 AngryPhonebook                        2828             2844       +16     +0.6%     0.99x (?)
  2 Array2D                               2938             2969       +31     +1.1%     0.99x (?)
  3 ArrayAppend                            718              726        +8     +1.1%     0.99x (?)
  4 ArrayAppendReserved                    499              500        +1     +0.2%     1.00x (?)
  5 ArrayInClass                            95               95        +0     +0.0%     1.00x
  6 ArrayLiteral                          1035             1033        -2     -0.2%     1.00x (?)
  7 ArrayOfGenericPOD                      209              209        +0     +0.0%     1.00x
  8 ArrayOfGenericRef                     3565             3568        +3     +0.1%     1.00x (?)
  9 ArrayOfPOD                             173              173        +0     +0.0%     1.00x
 10 ArrayOfRef                            3469             3538       +69     +2.0%     0.98x (?)
 11 ArraySubscript                        1457             1464        +7     +0.5%     1.00x (?)
 12 ArrayValueProp                           8                8        +0     +0.0%     1.00x
 13 ArrayValueProp2                          8                8        +0     +0.0%     1.00x
 14 ArrayValueProp3                          8                8        +0     +0.0%     1.00x
 15 ArrayValueProp4                          8                8        +0     +0.0%     1.00x
 16 BitCount                                17               17        +0     +0.0%     1.00x
 17 ByteSwap                                 1                1        +0     +0.0%     1.00x
 18 Calculator                              40               39        -1     -2.5%     1.03x
 19 CaptureProp                           5803             5848       +45     +0.8%     0.99x (?)
 20 Chars                                  875              853       -22     -2.5%     1.03x (?)
 21 ClassArrayGetter                        13               12        -1     -7.7%     1.08x (?)
 22 DeadArray                              174              174        +0     +0.0%     1.00x
 23 Dictionary                            1954             1967       +13     +0.7%     0.99x (?)
 24 Dictionary2                           3628             3605       -23     -0.6%     1.01x (?)
 25 Dictionary3                           1337             1352       +15     +1.1%     0.99x (?)
 26 DictionaryBridge                      5274             5151      -123     -2.3%     1.02x (?)
 27 DictionaryLiteral                     2916             2897       -19     -0.7%     1.01x (?)
 28 DictionaryRemove                      5386             5134      -252     -4.7%     1.05x (?)
 29 DictionarySwap                         983              987        +4     +0.4%     1.00x (?)
 30 ErrorHandling                         2826             2837       +11     +0.4%     1.00x (?)
 31 GlobalClass                              0                0        +0         *         *
 32 Hanoi                                 3617             3622        +5     +0.1%     1.00x (?)
 33 HashTest                              1767             1726       -41     -2.3%     1.02x
 34 Histogram                              757              755        -2     -0.3%     1.00x (?)
 35 Integrate                              256              257        +1     +0.4%     1.00x (?)
 36 Join                                   507              502        -5     -1.0%     1.01x (?)
 37 LinkedList                            9826             9813       -13     -0.1%     1.00x (?)
 38 MapReduce                              416              422        +6     +1.4%     0.99x (?)
 39 Memset                                 217              220        +3     +1.4%     0.99x (?)
 40 MonteCarloE                          10578            10487       -91     -0.9%     1.01x (?)
 41 MonteCarloPi                         43901            43682      -219     -0.5%     1.01x (?)
 42 NSDictionaryCastToSwift              11663            12116      +453     +3.9%     0.96x (?)
 43 NSError                                367              373        +6     +1.6%     0.98x (?)
 44 NSStringConversion                     641              627       -14     -2.2%     1.02x (?)
 45 NopDeinit                            35612            35727      +115     +0.3%     1.00x (?)
 46 ObjectAllocation                       229              230        +1     +0.4%     1.00x (?)
 47 OpenClose                               70               72        +2     +2.9%     0.97x
 48 Phonebook                            11072            12280     +1208    +10.9%     0.90x
 49 PolymorphicCalls                       103              105        +2     +1.9%     0.98x (?)
 50 PopFrontArray                         2011             1976       -35     -1.7%     1.02x (?)
 51 PopFrontArrayGeneric                  1887             1982       +95     +5.0%     0.95x (?)
 52 PopFrontUnsafePointer                 9058             8987       -71     -0.8%     1.01x (?)
 53 Prims                                  803              789       -14     -1.7%     1.02x (?)
 54 ProtocolDispatch                      3323             3187      -136     -4.1%     1.04x (?)
 55 ProtocolDispatch2                      146              150        +4     +2.7%     0.97x (?)
 56 RC4                                    227              230        +3     +1.3%     0.99x (?)
 57 RGBHistogram                          3078             3143       +65     +2.1%     0.98x (?)
 58 RangeAssignment                        337              336        -1     -0.3%     1.00x (?)
 59 RecursiveOwnedParameter               2006             2035       +29     +1.4%     0.99x (?)
 60 SetExclusiveOr                        3502             3450       -52     -1.5%     1.02x (?)
 61 SetIntersect                           405              410        +5     +1.2%     0.99x (?)
 62 SetIsSubsetOf                          582              583        +1     +0.2%     1.00x (?)
 63 SetUnion                              3251             3291       +40     +1.2%     0.99x (?)
 64 SevenBoom                             1403             1396        -7     -0.5%     1.01x (?)
 65 Sim2DArray                             841              813       -28     -3.3%     1.03x (?)
 66 SortLettersInPlace                    1036             1045        +9     +0.9%     0.99x (?)
 67 SortStrings                           1830             1814       -16     -0.9%     1.01x (?)
 68 StaticArray                           3080             3115       +35     +1.1%     0.99x (?)
 69 StrComplexWalk                        3447             3327      -120     -3.5%     1.04x (?)
 70 StrToInt                              4216             4408      +192     +4.6%     0.96x
 71 StringBuilder                         1288             1256       -32     -2.5%     1.03x (?)
 72 StringInterpolation                  11480            11706      +226     +2.0%     0.98x (?)
 73 StringWalk                            5109             5104        -5     -0.1%     1.00x (?)
 74 StringWithCString                    78312            83282     +4970     +6.3%     0.94x (?)
 75 SuperChars                          605642           603634     -2008     -0.3%     1.00x (?)
    Totals                              931614           935966     +4352     +0.5%     1.00x (?)
 76 TwoSum                                1807             1809        +2     +0.1%     1.00x (?)
 77 TypeFlood                                0                0        +0         *         *
 78 UTF8Decode                             916              918        +2     +0.2%     1.00x (?)
 79 Walsh                                  378              379        +1     +0.3%     1.00x (?)
 80 XorLoop                                369              370        +1     +0.3%     1.00x (?)
````

jckarter commented:
Does the additional force-inlining impact code size much?

lorentey commented:
Good question! Is there a specific statistic I should look at?

For what it's worth, `Benchmark_O` grows from 951568 to 956352; the delta is 4784 bytes, or about +0.5%. `libswiftCore.dylib` for x86-64 gets from 3783984 to 3790712 (6728 bytes, +0.18%).

jckarter commented:
Sounds acceptable to me (though that's @nadavrot's call ultimately).

dabrahams commented on stdlib/public/core/ArrayBufferType.swift:
```diff
@@ -157,28 +157,31 @@ extension _ArrayBufferType {
+      var generator = newValues.generate()
+      for i in subRange { 
+        let element = generator.next()
+        _collectionAssert(element != nil)
```
This assertion seems like it’s going to slow down non-optimized code just to get a better error message than the force-unwrap would supply.  Is it worth it?

lorentey replied:
I haven't realized this was going to have a meaningful effect for -Onone. The original code was happy to use an out-of-bounds index if the collection was broken, but I think the consistent error is desirable. However, if it causes a slowdown, I agree it should be removed.

I'll take a look at the unoptimized benchmarks; if removing them makes an impact, I'll do so.

dabrahams commented on stdlib/public/core/ArrayBufferType.swift:
```diff
@@ -157,28 +157,31 @@ extension _ArrayBufferType {
-        i._successorInPlace()
+      for i in oldTailIndex..<newTailIndex {
+        let element = generator.next()
+        _collectionAssert(element != nil)
```
Another potential slowdown for -Onone.  Why?

dabrahams commented on stdlib/public/core/ArrayBufferType.swift:
```diff
@@ -157,28 +157,31 @@ extension _ArrayBufferType {
+        (elements + i).initialize(element!)
       }
-      _expectEnd(i, newValues)
+      _collectionAssert(generator.next() == nil)
```
This one is okay because it doesn’t occur for every single element!

dabrahams commented on stdlib/public/core/ArrayBufferType.swift:
```diff
@@ -157,28 +157,31 @@ extension _ArrayBufferType {
       for _ in 0..<newCount {
-        elements[i] = newValues[j]
+        let element = generator.next()
+        _collectionAssert(element != nil)
```
Another slowdown for -Onone

dabrahams commented on stdlib/private/StdlibCollectionUnittest/LoggingWrappers.swift.gyb:
```diff
@@ -556,11 +556,15 @@ public struct Logging${Kind}<Base: ${Kind}Type> : ${Kind}Type, LoggingType {
-    -> UnsafeMutablePointer<Base.Generator.Element> {
+  /// Copy `self` into uninitialized contiguous memory of size `capacity`,
+  /// returning one past the last element initialized, or a `nil` pointer
+  /// if the copy cannot be safely made given `capacity`.
```
What does “safely made” mean here?  What does “given `capacity`” mean?

dabrahams commented on stdlib/private/StdlibCollectionUnittest/LoggingWrappers.swift.gyb:
```diff
@@ -556,11 +556,15 @@ public struct Logging${Kind}<Base: ${Kind}Type> : ${Kind}Type, LoggingType {
+  /// Copy `self` into uninitialized contiguous memory of size `capacity`,
+  /// returning one past the last element initialized, or a `nil` pointer
+  /// if the copy cannot be safely made given `capacity`.
+  public func _initializeTo(
```
I really don’t like the way uses of this new function signature reads.
Consider re-casting it [the way it will soon be renamed](https://github.com/apple/swift/blob/swift-3-api-guidelines/stdlib/public/core/Collection.swift#L594) anyway.  I don’t understand what `capacity` is supposed to be.  What am I allowed/required to pass here?

lorentey replied:
`capacity` is the number of uninitialized elements that the caller made available in the buffer that `ptr` points to; indeed, this should be spelled out in the doc comment. 

The parameter is necessary because `_initializeTo` is defined on `SequenceType`, where there is no guaranteed `count` — so the caller might not be able to allocate enough space to fit all elements. An alternative would be to trust that the caller allocates enough space for `underestimatedCount` elements, but that didn't feel like the right API to me. Having the capacity explicitly supplied is harder to get wrong, and (in theory) allows the function to work in cases when the underestimate was low but the caller happened to allocate more space anyway due to e.g. fitting storage size on an exponential growth curve.

My initial instinct was to move the definition down to `CollectionType`, but @gribozavr didn't like that:

>> Would you be against a patch that removed _initializeTo from SequenceType and moved it into CollectionType instead?

> I think that's a reasonable start, but not a perfect long-term solution.  The issue is that, Array() initializer takes a sequence (as do other functions that would use initalizeTo(), like appendContentsOf()).  We can add an overload for collections, but it is not a great solution.  Sometimes happens that a collection is statically known to be only a sequence (for example, when the Array() initializer is called from a generic function).  That would force collections to go through a slower code path.

> Another reasonable thing to do would be to allow _initializeTo to somehow refuse to operate, if it is not dynamically a collection, and then have a fallback on the caller side.  See _customContainsEquatableElement and how it is used for an example of this trick.

I think the argument for allowing `_initializeTo` to work in `SequenceType` context is a good one. I went with Dmitri's suggestion and allowed the method to fail; I added the capacity parameter during the course of the implementation.

* * *

The `_initializeTo` in this PR is somewhat limited by only allowing the sequence to be copied into a single contiguous buffer. This pretty much limits its use to `Array`. I'd actually prefer to see a non-failing but continuable variant that would allow its use in any collection that segments its storage into multiple contiguous regions (e.g., a deque based on a circular buffer, or even my B-tree collections). This would make the method signature even more complicated in `Sequence`, though. OTOH, it would fit a bit better into `IteratorProtocol`, with a declaration that could be a variant of this:

```
/// Copy at most `capacity` elements starting from the current element into
/// the buffer pointed to by `ptr`, advancing iteration state and the pointer itself.
/// - Returns: `true` if there are more elements remaining in the sequence.
mutating func _copyContents(to ptr: inout UnsafeMutablePointer<Element>, withCapacity capacity: Int) -> Bool
```

(I don't like the `inout` parameter or the boolean result though.) However, having `_copyContents` in `IteratorProtocol` would probably mean we'd need to change `Array` to use a specialized iterator other than `IndexingIterator`. This would mean writing up a public API proposal that'd be hard to justify unless `copyContents` was made public at the same time — and I'm not yet sure if the speedup would be significant enough to be worth that effort.

I'd  be OK with any of the following options:

1. Leave the `capacity` parameter in place (possibly with a better label), but explicitly document what it does.
2. Drop the parameter, but document that the caller is supposed to supply enough space to fit `underestimatedCount` elements. The method would still need to be able to fail, though.
3. Move the method down to `Collection`, restoring it to the original signature, but making it less useful.
4. Drop the `_initializeTo` part of this PR, and concentrate on replacing indexing with iterators (along with speeding up iterators when necessary).
5. Do more work to convert `_initializeTo` into a restartable/continuable thing, possibly moving it into `IteratorProtocol`.

Thoughts?

dabrahams replied:
on Mon Mar 07 2016, Károly Lőrentey <notifications-AT-github.com> wrote:

>> @@ -556,11 +556,15 @@ public struct Logging${Kind}<Base: ${Kind}Type> : ${Kind}Type, LoggingType {
>>      return base._copyToNativeArrayBuffer()
>>    }
>>  
>> -  /// Copy a Sequence into an array.
>> -  public func _initializeTo(ptr: UnsafeMutablePointer<Base.Generator.Element>)
>> -    -> UnsafeMutablePointer<Base.Generator.Element> {
>> +  /// Copy `self` into uninitialized contiguous memory of size `capacity`,
>> +  /// returning one past the last element initialized, or a `nil` pointer
>> +  /// if the copy cannot be safely made given `capacity`.
>> +  public func _initializeTo(
>
> `capacity` is the number of uninitialized elements that the caller
> made available in the buffer that `ptr` points to; indeed, this should
> be spelled out in the doc comment.

It doesn't read well at the call site because it isn't clear what it's
the capacity of.  Anywhere you are going to pass an
`UnsafeMutablePointer` + length of its potential storage, it might be
better to use an `UnsafeMutableBufferPointer`... although that was not
really designed to be used with uninitialized memory, you can always get
the baseAddress.

> The parameter is necessary because `_initializeTo` is defined on
> `SequenceType`, where there is no guaranteed `count` — so the caller
> might not be able to allocate enough space to fit all elements. An
> alternative would be to trust that the caller allocates enough space
> for `underestimatedCount` elements, but that didn't feel like the
> right API to me. Having the capacity explicitly supplied is harder to
> get wrong, and (in theory) allows the function to work in cases when
> the underestimate was low but the caller happened to allocate more
> space anyway due to e.g. fitting storage size on an exponential growth
> curve.

This makes sense to me... but only if we really need this method on
`SequenceType`.  Do we?  Otherwise it's premature generalization.

> The `_initializeTo` in this PR is somewhat limited by only allowing
> the sequence to be copied into a single contiguous buffer. This pretty
> much limits its use to `Array`. I'd actually prefer to see a
> non-failing but continuable variant that would allow its use in any
> collection that segments its storage into multiple contiguous regions
> (e.g., a deque based on a circular buffer, or even my B-tree
> collections). This would make the method signature even more
> complicated in `Sequence`, though. OTOH, it would fit a bit better
> into `IteratorProtocol`, with a declaration that could be a variant of
> this:
>
> ```
> /// Copy at most `capacity` elements starting from the current element into
> /// the buffer pointed to by `ptr`, advancing iteration state and the pointer itself.
> /// - Returns: `true` if there are more elements remaining in the sequence.
> mutating func _copyContents(to ptr: inout UnsafeMutablePointer<Element>, withCapacity capacity: Int) -> Bool
> ```

Why do you think this belongs in `IteratorProtocol`?

> (I don't like the `inout` parameter or the boolean result though.)

Me neither.  Why do it that way?

> However, having `_copyContents` in `IteratorProtocol` would probably
> mean we'd need to change `Array` to use a specialized iterator other
> than `IndexingIterator`. 

...because?

> This would mean writing up a public API proposal that'd be hard to
> justify unless `copyContents` was made public at the same time — and
> I'm not yet sure if the speedup would be significant enough to be
> worth that effort.
>
> I'd  be OK with any of the following options:
>
> 1. Leave the `capacity` parameter in place (possibly with a better label), but explicitly document what it does.
> 2. Drop the parameter, but document that the caller is supposed to
> supply enough space to fit `underestimatedCount` elements. The method
> would still need to be able to fail, though.
> 3. Move the method down to `Collection`, restoring it to the original
> signature, but making it less useful.

Presumably you'd need to do something else for `SequenceType` somewhere?
Otherwise, we'd be introducing complexity to the signature just for
theoretical utility with `SequenceType` that we're not actually taking
advantage of, which sounds like a bad idea.

> 4. Drop the `_initializeTo` part of this PR, and concentrate on
> replacing indexing with iterators (along with speeding up iterators
> when necessary).  5. Do more work to convert `_initializeTo` into a
> restartable/continuable thing, possibly moving it into
> `IteratorProtocol`.
>
> Thoughts?

IIUC the main point of this PR is that using iterators is better than
indexing; if that is separable from the `_initializeTo` change, I'd much
rather accept that part alone, and consider other changes in a different
PR.

-- 
-Dave

lorentey replied:
>>  `capacity` is the number of uninitialized elements that the caller
 made available in the buffer that `ptr` points to; indeed, this should
 be spelled out in the doc comment.

> It doesn't read well at the call site because it isn't clear what it's
the capacity of.  Anywhere you are going to pass an
`UnsafeMutablePointer` + length of its potential storage, it might be
better to use an `UnsafeMutableBufferPointer`... although that was not
really designed to be used with uninitialized memory, you can always get
the baseAddress.

The thing about that is that this would be an irregular use of `UnsafeMutableBufferPointer`: the implementation may *only* call `baseAddress` and `count`, as calling anything else would use uninitialized memory. This would deserve a warning in the docs, and we'd still be in a state where the call site looks confusing/wrong unless you read the documentation.

What if we replaced count with an end pointer? E.g. 
```
let end = copyContents(to: ptr, limit: ptr + capacity)
```
(`ForwardIndex.advanced(by:limit:)` uses a `limit` label in a similar context.)

>>  ```
>> /// Copy at most `capacity` elements starting from the current element into
>> /// the buffer pointed to by `ptr`, advancing iteration state and the pointer itself.
>> /// - Returns: `true` if there are more elements remaining in the sequence.
>> mutating func _copyContents(to ptr: inout UnsafeMutablePointer<Element>, withCapacity capacity: Int) -> Bool
>>  ```

> Why do you think this belongs in `IteratorProtocol`?

If we want to be able to copy elements of a sequence into a series of segments, an iterator would be the natural way to represent the progress of copying. So the copy method (wherever it is defined) would need an inout iterator to record progress, beside the parameter(s) that describe the buffer to use (i.e., a base address + length, in some form). It needs to return the number of elements copied, and a flag that indicates whether or not it reached the end of the sequence.

Putting the copy method inside `IteratorProtocol` would get rid of the inout iterator. Also, iterators already contain enough knowledge about the sequence to retrieve a single element, so it doesn't seem too far-fetched that they would also be able to copy multiple elements at once.

>> (I don't like the `inout` parameter or the boolean result though.)

> Me neither.  Why do it that way?

I am yet to find a variant that I'm entirely happy with. Perhaps this would be better:
```
mutating func _copyContents(to bufferPointer: UMP, limit: UMP) -> (end: UMP, finished: Bool)
```
My difficulty with coming up with a really nice signature is the reason I haven't followed up on this idea yet. It may indicate that the proposed concept is just too complex, but I don't think I mastered the naming conventions to a level where I could say for sure.

>> However, having `_copyContents` in `IteratorProtocol` would probably
 mean we'd need to change `Array` to use a specialized iterator other
 than `IndexingIterator`.

> ...because?

Because I don't see how we could implement `Array`'s `_copyContents` in `IndexingIterator`.

Unless we introduced an extra hook in `Collection` that `IndexingIterator` could call into? Hm. Since that hook would be in `Collection` rather than `Sequence`, it would be quite a bit simpler… This could be interesting.

> IIUC the main point of this PR is that using iterators is better than
indexing; if that is separable from the `_initializeTo` change, I'd much
rather accept that part alone, and consider other changes in a different
PR.

Wholeheartedly agreed! I'll remove the `_initializeTo` stuff from this PR and reintroduce it after a little more time in the oven.

rudkx commented:
@lorentey Is this something you might consider rebasing or reworking on ToT?

lorentey commented:
@rudkx Yes, of course, I'll have a go at it. I'll go with @dabrahams's suggestion and reduce the scope to preferring iterators over indices — which seems to be a straightforward enough change.

Then I'll see if I can convert _initializeTo into something more flexible, and submit a new PR if I get somewhere interesting.
